### PR TITLE
Add cardholder edit v1

### LINKED ISSUE #154: Implement cardholder edit frontend + functionality
Allow cardholders to edit meaning, telegram, and website.

### LINKED ISSUE #159: Add quest ideation form to cardholder page
We need to have an intake form for quest ideas, and also set up email notifications when it gets filled out.

### PR DESCRIPTION
**Pull request by vb7401**
Closes #154 
Closes #159

### CODE REVIEW

File: apps/jubmoji-quest/next.config.js [hunk 1]
```diff
@@ -35,6 +35,11 @@ const nextConfig = {
         destination: "https://airtable.com/apprikYUs5WuXJRbN/shrIHSLl3CiJ76OxI",
         permanent: true,
       },
+      {
+        source: "/quest-proposal",
```
Comment (AndrewCLu) [inline apps/jubmoji-quest/next.config.js:39]: nit: `/quest-form` might be a bit easier to remember

File: apps/jubmoji-quest/src/pages/cardholder/index.tsx [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+  NfcCardRawSignature,
+} from "jubmoji-api";
+import { bigIntToHex } from "babyjubjub-ecdsa";
+import Link from "next/link";
+import { Textarea } from "@/components/ui/Textarea";
+
+enum CardholderActions {
+  EDIT_CARD_DETAILS,
+  PROPOSE_QUEST,
+  PSEUD_X_POST,
```
Comment (AndrewCLu) [inline apps/jubmoji-quest/src/pages/cardholder/index.tsx:30]: X...disgusting

File: apps/jubmoji-quest/src/pages/cardholder/index.tsx [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+  UNLOCKING,
+  EDITING,
+  CONFIRMING,
+}
+
+interface CardholderTapModalProps extends ModalProps {
+  action: CardholderActions;
+}
+
+const CardholderTapModal = ({
```
Comment (AndrewCLu) [inline apps/jubmoji-quest/src/pages/cardholder/index.tsx:193]: This should be refactored into modals

File: apps/jubmoji-quest/src/pages/cardholder/index.tsx [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+        },
+      });
+
+      const proofInstance = createProofInstance(PublicMessageSignature, {
+        randStr: "",
+      });
+      const pubKeyIndex = cardPubKeys.findIndex(
+        (key) => key.pubKeySlot1 === res.publicKey // Use secp256k1 key here, which is in slot 1
+      );
+      const proof = await proofInstance.prove({
```
Comment (AndrewCLu) [inline apps/jubmoji-quest/src/pages/cardholder/index.tsx:246]: Why do we need to prove and verify here?

File: apps/jubmoji-quest/src/pages/cardholder/index.tsx [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+
+  const [modalState, setModalState] = useState(
+    CardholderTapModalState.UNLOCKING
+  );
+  const [pubKeyIndex, setPubKeyIndex] = useState<number>();
+  const [meaning, setMeaning] = useState<string>();
+  const [telegramLink, setTelegramLink] = useState<string>();
+  const [website, setWebsite] = useState<string>("");
+
+  const initialTap = async () => {
```
Comment (AndrewCLu) [inline apps/jubmoji-quest/src/pages/cardholder/index.tsx:214]: Lots of shared code with `ForegroundTapModal` here. We can duplicate this for now to unblock, but should track this somewhere.

File: apps/jubmoji-quest/src/pages/cardholder/index.tsx [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+      } else {
+        toast.error("Cardholder verification failed, please try again.");
+      }
+    } catch (error) {
+      console.error(error);
+      toast.error("Tapping failed, please try again.");
+    }
+  };
+
+  const sendTap = async () => {
```
Comment (AndrewCLu) [inline apps/jubmoji-quest/src/pages/cardholder/index.tsx:272]: Some comment about code reuse here.

File: apps/jubmoji-quest/src/pages/cardholder/index.tsx [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+      });
+      const { verified } = await proofInstance.verify(proof);
+
+      if (verified) {
+        const response = await fetch("/api/updateCard", {
+          method: "POST",
+          headers: {
+            "Content-Type": "application/json",
+          },
+          body: JSON.stringify({
```
Comment (AndrewCLu) [inline apps/jubmoji-quest/src/pages/cardholder/index.tsx:326]: Should just send the proof to the backend here.

File: apps/jubmoji-quest/src/pages/cardholder/index.tsx [hunk 8]
```diff
@@ -0,0 +1,486 @@
...
+  {
+    title: string;
+    buttonText: string;
+    entryFields?: { field: string; size: "big" | "small" }[];
+    redirect?: string;
+  }
+> = {
+  [CardholderActions.EDIT_CARD_DETAILS]: {
+    title: "Cardholder edit portal",
+    buttonText: "Unlock editing",
```
Comment (AndrewCLu) [inline apps/jubmoji-quest/src/pages/cardholder/index.tsx:46]: I don't think these should say "Unlock" - the phrasing is quite confusing. Think its better to just use directives such as "Begin Editing"

File: apps/jubmoji-quest/src/pages/cardholder/index.tsx [hunk 9]
```diff
@@ -0,0 +1,486 @@
...
+                backgroundRepeat: "no-repeat",
+              }}
+            ></div>
+
+            <span className=" font-dm-sans text-[13px] font-semibold text-shark-50">
+              Meaning*
+            </span>
+            <Textarea
+              onChange={(event) => setMeaning(event.target.value)}
+              placeholder="Edit your jubmoji's meaning"
```
Comment (AndrewCLu) [inline apps/jubmoji-quest/src/pages/cardholder/index.tsx:410]: These should all prefill the existing values

File: apps/jubmoji-quest/src/pages/cardholder/index.tsx [hunk 10]
```diff
@@ -0,0 +1,486 @@
...
+              rows={1}
+            />
+
+            <span className=" font-dm-sans text-[13px] font-semibold text-shark-50">
+              Telegram link*
+            </span>
+            <Textarea
+              onChange={(event) => setTelegramLink(event.target.value)}
+              placeholder="t.me/username or chat link"
```
Comment (AndrewCLu) [inline apps/jubmoji-quest/src/pages/cardholder/index.tsx:419]: Should add t.me prefix automatically

File: apps/jubmoji-quest/src/pages/cardholder/index.tsx [hunk 11]
```diff
@@ -0,0 +1,486 @@
...
+                  cardPubKeys[pubKeyIndex!].imageBlobUrl
+                })`,
+                backgroundSize: "cover",
+                backgroundPosition: "center",
+                backgroundRepeat: "no-repeat",
+              }}
+            ></div>
+
+            <span className=" font-dm-sans text-[13px] font-semibold text-shark-50">
+              Meaning*
```
Comment (AndrewCLu) [inline apps/jubmoji-quest/src/pages/cardholder/index.tsx:406]: Shouldn't be required to update all fields

**Review by AndrewCLu — APPROVED** [review]
Unblocking this for now, but there are quite a few polishes/refactors that we should make sure are tracked.

File: apps/jubmoji-quest/src/pages/cardholder/index.tsx (hunk 2 above)
Comment (vb7401) [reply apps/jubmoji-quest/src/pages/cardholder/index.tsx:30]: LOL

File: apps/jubmoji-quest/src/pages/cardholder/index.tsx (hunk 4 above)
Comment (vb7401) [reply apps/jubmoji-quest/src/pages/cardholder/index.tsx:246]: Idk just to do some error handling, but ill remove

File: apps/jubmoji-quest/src/pages/cardholder/index.tsx (hunk 8 above)
Comment (vb7401) [reply apps/jubmoji-quest/src/pages/cardholder/index.tsx]: good point fixed!

**Pull request merged**